test/ConfCrypt/Commands/Tests.hs

```diff
@@ -84,23 +84,42 @@ readTests = testGroup "Read" [
 
    ,testCase "reading an empty file is an empty file" $ do
         let testLines = parseConfCrypt "empty test file" "# just a comment"
-        res <- getReadResult testLines
+        res <- getReadResult Nothing testLines
         case res of
             Left e ->
                 assertFailure $ show e
             Right lines ->
                 lines @=? []
 
+    ,testCase "user-specified formats are used to render output" $ do
+        let testLines = parseConfCrypt "test file" testFile
+        res <- getReadResult (Just "foo%t %n=%v %%") testLines
+        case res of
+            Left e ->
+                assertFailure $ show e
+            Right lines ->
+                lines @=? ["fooString Test=Foobar %", "fooInt Test2=42 %"]
+
+    ,testCase "invalid formats give an error" $ do
+        let testLines = parseConfCrypt "test file" testFile
+        res <- getReadResult (Just "foo%abar") testLines
+        case res of
+            Left e ->
+                e @=? FormatParseError "1:6:\n  |\n1 | foo%abar\n  |      ^\nUnrecognized variable a\n"
+                -- This is ugly, but it can't be improved without revamping error message formatting in general
```

Comment:
As far as I can see, this can be trivially done by providing a custom `Show` instance for `ConfCryptError`:

```
instance Show ConfCryptError where
    show (FormatParseError err) -> "Error parsing format: " <> err
    show other = gShow other
```

Or, better yet, provide sensible formatting for every option, don't resort to `gShow`.

Reply:
I think this problem goes a little deeper, which is why I opted not to solve it here. For example, if you pass a malformed file to confcrypt, it parses up to the invalid line and returns the results it parsed, but doesn't return any error and exits with code 0.

Reply:
I see

Reply:
Soo, I just verified that this does in fact happen. I want to fix the underlying issue before merging this.

Its basically because the file format is parsed as a list using `many`. Its pretty easy to fix, I think.

Reply:
Okay, the parser has been patched. Please rebase & update your test. I'd love to get this merged this week.

Reply:
Fixing that issue didn't change the result of my test, it still passed. The parser bailing out early was just an example of another error handling issue, the core issue is that confcrypt `show`s errors rather than using the appropriate formatting function.

Another example: try changing the `String` in one of the types in a confcrypt file to `str`. The resulting error message is 11136 characters long and contains most of the text of the file.

Reply:
I think there were two distinct issues. To me, the parser not short-circuiting was actually a bug. The error text was just terrible UX. 

I've fixed the parser error text though, so it pretty prints now.

Reply:
Turns out I didn't need to update the test at all!